src/coreclr/jit/flowgraph.cpp

```diff
@@ -52,6 +52,244 @@ static bool blockNeedsGCPoll(BasicBlock* block)
+                            break;
+                        }
+                    }
+                    if (hasNotSupportedPreds)
+                    {
+                        continue;
+                    }
+
+                    if (prevBb == nullptr)
+                    {
+                        // We're going to emit a BB in front of fgFirstBB
+                        fgEnsureFirstBBisScratch();
+                        prevBb = fgFirstBB;
+                        if (prevBb == block)
+                        {
+                            continue;
+                        }
+                    }
+
+                    // So, we found a helper call inside the "block" - let's extract it to a
+                    // separate block "callInitBb" and guard it with a fast "isInitedBb" bb.
+                    // The final layout should look like this:
+
+                    // BB0 "prevBb":
+                    //     ...
+                    //
+                    // BB1 "isInitedBb":    (preds: BB0 + %current preds of BB3%)
+                    //
+                    //  *  JTRUE     void
+                    //  \--*  NE        int
+                    //     +--*  AND       int
+                    //     |  +--*  IND       ubyte
+                    //     |  |  \--*  CNS_INT   long   moduleIdArg + clsIdArg + dataBlobOffset
+                    //     |  \--*  CNS_INT   int    isInitMask
+                    //     \--*  CNS_INT   int    0
+                    //
+                    //
+                    // BB2 "callInitBb":    (preds: BB1)
+                    //
+                    //  *  CALL help long   HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
+                    //  +--*  CNS_INT   long   moduleIdArg
+                    //  \--*  CNS_INT   int    clsIdArg
+                    //
+                    // BB3 "block"          (preds: BB1, BB2)
+                    //     ...
+                    //
+
+                    // TODO: ask VM for these constants:
+                    const int dataBlobOffset = 48; // DomainLocalModule::GetOffsetOfDataBlob()
+                    const int isInitMask     = 1;  // ClassInitFlags::INITIALIZED_FLAG;
+
+                    UINT8* isInitAdr = (UINT8*)moduleIdArg->AsIntCon()->IconValue() + dataBlobOffset +
+                                       clsIdArg->AsIntCon()->IconValue();
+
+                    GenTree* isInitAdrNode =
+                        gtNewIndOfIconHandleNode(TYP_UBYTE, (size_t)isInitAdr, GTF_ICON_CONST_PTR, true);
+
+
+                    // Let's start from emitting that BB2 "callInitBb"
+                    BasicBlock* callInitBb = fgNewBBbefore(BBJ_NONE, block, true);
+                    // it's executed only once so can be marked as cold
+                    callInitBb->bbSetRunRarely();
+                    callInitBb->bbFlags |= (BBF_INTERNAL | BBF_HAS_CALL | BBF_HAS_LABEL);
+                    GenTree* clonedHelperCall = gtCloneExprCallHelper(call);
```

Comment:
Do you need to clone the call? Can't you just move the tree to a new stmt?

Reply:
Changed the code to emit a new call there now (`CORINFO_HELP_INITCLASS`) which is slower but takes just a single argument instead of two (smaller)